### ISSUE TITLE
Default ConfigSource for file system properties

### ISSUE THREAD

**Issue by sdaschner** [eclipse/microprofile-config#316]
A commonly required use case for container orchestration environments (Kubernetes) is to map files into the container's file system (K8s does this, e.g. using _config maps_).

It would be very helpful to include a default `ConfigSource` that loads a properties file from a default location, such as `/opt/config/microprofile-config.properties`, or `~/microprofile-config.properties`, etc.

Have a look at ConfigJSR/issues/7 in the Config JSR.

In a previous blog post I showed how to do this with classpath resource properties files -- that would be possible today with MP Config -- but it's more effort in a few applications servers: https://blog.sebastian-daschner.com/entries/kubernetes_configmaps_javaee

**Comment by Emily-Jiang** [comment]
It is useful to have this kind of config source for K8s. There are a couple of options:

1. microservice code provides this config source by implementing ConfigSource api
2. app server can provide this config source to apps 
3. code template for creating microservices for K8s can automatically create the custom config source and utlise MP Config feature.

I think 2 or 3 is better than 1. Baking the environment related requirements in the app is not a clean design, as the app might be deployed in different environment though.

**Comment by sdaschner** [comment]
Yes, I agree. You're right that this can be done by the app server, however, developer would like a vendor-independent way. I think it's a very common approach that makes sense to be added to MP Config in general, i.e. requiring vendors to provide that default ConfigSource (similar to our discussions in the Config JSR).

**Comment by OndrejM** [comment]
I agree that this is very useful. In fact, WF Swarm provides and very soon also Payara will provide such a config source.

It makes sense to require such a config source from all MP implementations, but we should clarify how to ensure this. Providing a config source isn't related to the API, so a common spec could only require that:

- a directory based config source exists and supports a specified directory structure
- a standard location of such a directory - I think this is complicated because it's environment-specific (a different location makes sense on Linux, Windows and other systems)
- a standard way to configure the location of such a directory - we could specify a system property and env variable. We could also specify an MP Config property but then MP Config would be used to configure itself and it brings complications. Another complication would be if we wanted to support more directories.

So in the end, we could separate this to 2 things
 - specify the format of a directory for such a config source, its default ordinal and require that it's implemented, including TCK tests
- specify a common way to configure the location of the directory/directories or specify a default location - this may be problematic and maybe even unnecessary as each implementation will most probably enable to configure it once it has to provide the config source

**Comment by arthurdm** [comment]
+1 to this request.  It is very useful in a k8s world to have a folder in the container where you can mount secrets into property files and have those available to MP Config variables in the code.  

Passing secrets via env variables is not a secure solution, so we end up not having a vendor-neutral OOTB solution for the volume mount.
